// ng2-bootstrap/components/buttons/button-radio.js
System.register(['angular2/core', 'angular2/common'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var core_1, common_1;
    var ButtonRadio;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (common_1_1) {
                common_1 = common_1_1;
            }],
        execute: function() {
            ButtonRadio = (function () {
                function ButtonRadio(cd, el) {
                    this.cd = cd;
                    this.el = el;
                    this.onChange = function (_) { };
                    this.onTouched = function () { };
                    // hack!
                    cd.valueAccessor = this;
                }
                Object.defineProperty(ButtonRadio.prototype, "isActive", {
                    get: function () {
                        return this.btnRadio === this.value;
                    },
                    enumerable: true,
                    configurable: true
                });
                ButtonRadio.prototype.onClick = function () {
                    if (this.uncheckable && this.btnRadio === this.value) {
                        return this.cd.viewToModelUpdate(null);
                    }
                    this.cd.viewToModelUpdate(this.btnRadio);
                };
                ButtonRadio.prototype.ngOnInit = function () {
                    this.uncheckable = typeof this.uncheckable !== 'undefined';
                };
                Object.defineProperty(ButtonRadio.prototype, "value", {
                    // hack view model!
                    get: function () {
                        return this.cd.viewModel;
                    },
                    set: function (value) {
                        this.cd.viewModel = value;
                    },
                    enumerable: true,
                    configurable: true
                });
                // ControlValueAccessor
                // model -> view
                ButtonRadio.prototype.writeValue = function (value) {
                    this.value = value;
                };
                ButtonRadio.prototype.registerOnChange = function (fn) {
                    this.onChange = fn;
                };
                ButtonRadio.prototype.registerOnTouched = function (fn) {
                    this.onTouched = fn;
                };
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', String)
                ], ButtonRadio.prototype, "btnRadio", void 0);
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', Boolean)
                ], ButtonRadio.prototype, "uncheckable", void 0);
                __decorate([
                    core_1.HostBinding('class.active'), 
                    __metadata('design:type', Object)
                ], ButtonRadio.prototype, "isActive", null);
                __decorate([
                    core_1.HostListener('click'), 
                    __metadata('design:type', Function), 
                    __metadata('design:paramtypes', []), 
                    __metadata('design:returntype', void 0)
                ], ButtonRadio.prototype, "onClick", null);
                ButtonRadio = __decorate([
                    core_1.Directive({ selector: '[btnRadio][ngModel]' }),
                    __param(0, core_1.Self()), 
                    __metadata('design:paramtypes', [common_1.NgModel, core_1.ElementRef])
                ], ButtonRadio);
                return ButtonRadio;
            })();
            exports_1("ButtonRadio", ButtonRadio);
        }
    }
});
