// ng2-bootstrap/components/collapse/collapse.js
System.register(['angular2/core'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var core_1;
    var Collapse;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            }],
        execute: function() {
            // todo: add animate
            // todo: add init and on change
            Collapse = (function () {
                function Collapse() {
                    // shown
                    this.isExpanded = true;
                    // hidden
                    this.isCollapsed = false;
                    // stale state
                    this.isCollapse = true;
                    // animation state
                    this.isCollapsing = false;
                }
                Object.defineProperty(Collapse.prototype, "collapse", {
                    get: function () {
                        return this.isExpanded;
                    },
                    set: function (value) {
                        this.isExpanded = value;
                        this.toggle();
                    },
                    enumerable: true,
                    configurable: true
                });
                Collapse.prototype.toggle = function () {
                    if (this.isExpanded) {
                        this.hide();
                    }
                    else {
                        this.show();
                    }
                };
                Collapse.prototype.hide = function () {
                    var _this = this;
                    this.isCollapse = false;
                    this.isCollapsing = true;
                    this.isExpanded = false;
                    this.isCollapsed = true;
                    setTimeout(function () {
                        _this.height = '0';
                        _this.isCollapse = true;
                        _this.isCollapsing = false;
                    }, 4);
                };
                Collapse.prototype.show = function () {
                    var _this = this;
                    this.isCollapse = false;
                    this.isCollapsing = true;
                    this.isExpanded = true;
                    this.isCollapsed = false;
                    setTimeout(function () {
                        _this.height = 'auto';
                        _this.isCollapse = true;
                        _this.isCollapsing = false;
                    }, 4);
                };
                __decorate([
                    core_1.HostBinding('style.height'), 
                    __metadata('design:type', String)
                ], Collapse.prototype, "height", void 0);
                __decorate([
                    core_1.HostBinding('class.in'),
                    core_1.HostBinding('attr.aria-expanded'), 
                    __metadata('design:type', Boolean)
                ], Collapse.prototype, "isExpanded", void 0);
                __decorate([
                    core_1.HostBinding('attr.aria-hidden'), 
                    __metadata('design:type', Boolean)
                ], Collapse.prototype, "isCollapsed", void 0);
                __decorate([
                    core_1.HostBinding('class.collapse'), 
                    __metadata('design:type', Boolean)
                ], Collapse.prototype, "isCollapse", void 0);
                __decorate([
                    core_1.HostBinding('class.collapsing'), 
                    __metadata('design:type', Boolean)
                ], Collapse.prototype, "isCollapsing", void 0);
                __decorate([
                    core_1.Input(), 
                    __metadata('design:type', Boolean), 
                    __metadata('design:paramtypes', [Boolean])
                ], Collapse.prototype, "collapse", null);
                Collapse = __decorate([
                    core_1.Directive({ selector: '[collapse]' }), 
                    __metadata('design:paramtypes', [])
                ], Collapse);
                return Collapse;
            })();
            exports_1("Collapse", Collapse);
        }
    }
});
