// ng2-bootstrap/components/dropdown/dropdown-toggle.js
System.register(['angular2/core', './dropdown'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var core_1, dropdown_1;
    var DropdownToggle;
    return {
        setters:[
            function (core_1_1) {
                core_1 = core_1_1;
            },
            function (dropdown_1_1) {
                dropdown_1 = dropdown_1_1;
            }],
        execute: function() {
            DropdownToggle = (function () {
                function DropdownToggle(dropdown, el) {
                    this.dropdown = dropdown;
                    this.el = el;
                    this.disabled = false;
                    this.addClass = true;
                }
                DropdownToggle.prototype.ngOnInit = function () {
                    this.dropdown.dropDownToggle = this;
                };
                Object.defineProperty(DropdownToggle.prototype, "isOpen", {
                    get: function () {
                        return this.dropdown.isOpen;
                    },
                    enumerable: true,
                    configurable: true
                });
                DropdownToggle.prototype.toggleDropdown = function (event) {
                    event.stopPropagation();
                    if (!this.disabled) {
                        this.dropdown.toggle();
                    }
                    return false;
                };
                __decorate([
                    core_1.HostBinding('class.disabled'),
                    core_1.Input(), 
                    __metadata('design:type', Boolean)
                ], DropdownToggle.prototype, "disabled", void 0);
                __decorate([
                    core_1.HostBinding('class.dropdown-toggle'),
                    core_1.HostBinding('attr.aria-haspopup'), 
                    __metadata('design:type', Object)
                ], DropdownToggle.prototype, "addClass", void 0);
                __decorate([
                    core_1.HostBinding('attr.aria-expanded'), 
                    __metadata('design:type', Object)
                ], DropdownToggle.prototype, "isOpen", null);
                __decorate([
                    core_1.HostListener('click', ['$event']), 
                    __metadata('design:type', Function), 
                    __metadata('design:paramtypes', [MouseEvent]), 
                    __metadata('design:returntype', void 0)
                ], DropdownToggle.prototype, "toggleDropdown", null);
                DropdownToggle = __decorate([
                    core_1.Directive({ selector: '[dropdown-toggle]' }),
                    __param(0, core_1.Host()), 
                    __metadata('design:paramtypes', [dropdown_1.Dropdown, core_1.ElementRef])
                ], DropdownToggle);
                return DropdownToggle;
            })();
            exports_1("DropdownToggle", DropdownToggle);
        }
    }
});
